### PR TITLE
fix: implement DELTA learning rule for field coupling

### LINKED ISSUE #38: [Bug] DELTA learning rule is unimplemented (silent no-op)
**Environment**

| | |
|---|---|
| OS | All platforms (found by code inspection) |
| Compiler | Any |
| Version / commit | v2.7.0 |

**Steps to reproduce**

1. Create a `FieldCoupling` element with `FieldCouplingParameters` where `learningRule = LearningRule::DELTA` and `isLearningActive = true`.
2. Connect an input and output field and run the simulation so that the learning update path executes.
3. Observe that no weights are ever updated.

**Expected behaviour**

The Delta (error-driven) learning rule updates the coupling weight matrix each learning step, as advertised by the public `LearningRule::DELTA` enum value and its presence in `LearningRuleToString`.

**Actual behaviour**

The `DELTA` case is a stub: it only logs an error and breaks; the actual update call is commented out and the helper `tools::math::unsupervisedDeltaLearningRule` does not exist. Weights never change. `HEBB` and `OJA` work.

```cpp
// src/elements/field_coupling.cpp:180-183
case LearningRule::DELTA:
    log(tools::logger::LogLevel::ERROR, "Unsupervised delta learning rule is not implemented yet.");
    //tools::math::unsupervisedDeltaLearningRule(components["weights"], inputActivation, outputActivation, parameters.learningRate);
    break;
```

**Relevant log output**

```
[ERROR] Unsupervised delta learning rule is not implemented yet.
```

**Possible fix**

Implement `unsupervisedDeltaLearningRule` in `tools/math` (Δw = η · (target − output) · inputᵀ) and wire it into the `DELTA` case. **Note:** this bug blocks the planned "Delta learning rule example" (`examples/delta_learning.cpp`) — the example cannot exist until this rule works. Either expose a teacher/target signal or document the supervised target source.

### PR DESCRIPTION
**Pull request by Jgocunha**
Closes #38.

## Problem
`LearningRule::DELTA` for `FieldCoupling` was an unimplemented stub — the `DELTA` case only logged an error and broke; the real update was commented out and the helper `tools::math::unsupervisedDeltaLearningRule` did not exist. Weights never changed, unlike `HEBB`/`OJA`.

## Fix
- Implement `tools::math::unsupervisedDeltaLearningRule` (`include/tools/math.h`): the classic error-driven delta/Widrow-Hoff update `Δw[i][j] = η·(target[j] − predicted[j])·input[i]`, where `predicted[j] = Σᵢ w[i·outputSize+j]·input[i]`. Same flattened row-major `(inputSize×outputSize)` layout, `i*outputSize+j` indexing, and empty/size-mismatch guards as `hebbLearningRule`/`ojaLearningRule`.
- Wire it into the `DELTA` case in `src/elements/field_coupling.cpp`, replacing the error-log stub (matches the `HEBB`/`OJA` cases).

**Target/teacher semantics:** `FieldCoupling` only ever exposes the two normalized field activations to its learning rules — there is no separate externally-supplied teacher. To keep the rule genuinely error-driven (not degenerate Hebbian correlation) while staying unsupervised, the output field's own activation plays the target role (as it already does as the post-synaptic term for HEBB/OJA), corrected against the coupling's current linear estimate `wᵀ·input`. This reduces to the standard supervised delta rule the moment a caller supplies `output` as a genuine target. Rationale is documented above the function.

## Tests
- `tests/tools/test_math.cpp` — 5 tests: hand-computed update from zero weights, zero-error ⇒ no change, positive-error increases weight, empty-input throws, size-mismatch throws.
- `tests/elements/test_field_coupling.cpp` — 2 tests: zero-error produces no weight change; non-uniform target increases weights from zero.

## Verification
VS2022 + `VCPKG_ROOT=C:\dev-files\vcpkg`, Ninja/Release. Clean build; `*Delta*` filter 17/17; full suite **979/979** passing, no regressions.

## Summary by CodeRabbit

* **New Features**
  * Added unsupervised DELTA learning for field couplings, plus standard-normal sampling utilities and wrapped separable 2D kernel helpers.
* **Bug Fixes**
  * Enabled previously unavailable DELTA behavior in field coupling weight updates.
* **Performance Improvements**
  * Improved 1D/2D separable convolution execution with better optimized data handling.
* **Tests**
  * Expanded coverage for learning-rule outcomes (including invalid input cases), normal sampling statistics/reproducibility, and separable convolution/kernel correctness (including edge cases).

### CODE REVIEW

File: dynamic-neural-field-composer/src/elements/field_coupling.cpp [hunk 1]
```diff
@@ -180,8 +180,7 @@
 			switch (parameters.learningRule)
 			{
 			case LearningRule::DELTA:
-				log(tools::logger::LogLevel::ERROR, "Unsupervised delta learning rule is not implemented yet.");
-				//tools::math::unsupervisedDeltaLearningRule(components["weights"], inputActivation, outputActivation, parameters.learningRate);
+				tools::math::unsupervisedDeltaLearningRule(components["weights"], inputActivation, outputActivation, parameters.learningRate);
```
[the remark this answers is not in the document]
Comment (Jgocunha) [reply dynamic-neural-field-composer/src/elements/field_coupling.cpp]: Good catch — confirmed. `updateOutput()` applies `parameters.scalar` (`output = scalar * W * input`), so learning in the unscaled `W * input` space converged the weights to the wrong value whenever `scalar != 1`.

Fixed in `385db7bc`:
- `field_coupling.cpp` DELTA case now feeds a scalar-adjusted copy of the input to the helper, so both the prediction and the gradient carry the coupling scalar (learns the same scaled forward model `updateOutput()` evaluates).
- Revised the "same product" note in `math.h` to state that the caller passes scalar-adjusted input.
- Added `FieldCouplingDeltaLearningRule.NonUnitScalarZeroErrorLeavesNonzeroWeightsUnchanged`: with `scalar = 2` and hand-built nonzero weights whose *scaled* prediction reproduces the normalized target exactly (zero error), the weights must stay put — this fails on the pre-fix code and passes now.

Full suite 980/980 green.
